src/storage/buffer_manager.cpp

```diff
@@ -283,7 +322,24 @@ void BufferManager::AddToEvictionQueue(shared_ptr<BlockHandle> &handle) {
 	D_ASSERT(handle->readers == 0);
 	handle->eviction_timestamp++;
 	queue->q.enqueue(make_unique<BufferEvictionNode>(weak_ptr<BlockHandle>(handle), handle->eviction_timestamp));
-	// FIXME: do some house-keeping to prevent the queue from being flooded with many old blocks
+	PurgeQueue();
```

Comment:
There are 2 exit conditions to the loop in `PurgeQueue`. By purging nodes every time we add a node, the queue will stay clean, and we will rarely purge 1024 nodes. Although perhaps 1024 is excessive, this could be 32 or so.

I profiled this and it did not seem like it took much time at all. Purging is cheap since we are not flushing to disk, but just removing the redundant nodes.

Reply:
Ah, I see, the loop is exited once we encounter a single node that cannot be erased from the queue. The way I read it was that it would check the first 1024 nodes and evict any of the 1024 that are no longer relevant while re-queueing any that were, which seemed excessive to me. The initial implementation seems fine to me then actually.

Actually I would change it so that it is not even limited  1024 is not even there; just erase entries from the queue until you encounter a node that you cannot remove, and then stop.

Reply:
Alright, I'll change it back and make it a `while (true)`!